packages/ui/src/components/MessageInput/InputToolbar.tsx

```diff
@@ -0,0 +1,177 @@
+    const { isKeyboardUp } = useKeyboard();
+    const [toolbarContext, setToolbarContext] = React.useState<ToolbarContext>(
+      ToolbarContext.Main
+    );
+
+    const hideToolbar =
+      hidden === undefined ? !isKeyboardUp || !editorState.isFocused : hidden;
+
+    const itemIsActive = useCallback(
+      (item: ToolbarItem) => {
+        const args = {
+          editor,
+          editorState,
+          setToolbarContext,
+          toolbarContext,
+        };
+
+        return item.active(args);
+      },
+      [editor, editorState, setToolbarContext, toolbarContext]
+    );
+
+    const itemIsDisabled = useCallback(
+      (item: ToolbarItem) => {
+        const args = {
+          editor,
+          editorState,
+          setToolbarContext,
+          toolbarContext,
+        };
+
+        return item.disabled(args);
+      },
+      [editor, editorState, setToolbarContext, toolbarContext]
+    );
+
+    const createItemViewStyle = useCallback(
+      (item: ToolbarItem) => [
+        editor.theme.toolbar.toolbarButton,
+        itemIsActive(item) ? editor.theme.toolbar.iconWrapperActive : undefined,
+        itemIsDisabled(item)
+          ? editor.theme.toolbar.iconWrapperDisabled
+          : undefined,
+      ],
+      [editor, itemIsActive, itemIsDisabled]
+    );
+
+    const touchableStyle = useMemo(
+      () => editor.theme.toolbar.toolbarButton,
+      [editor.theme.toolbar.toolbarButton]
+    );
+
+    const renderItem: ListRenderItem<ToolbarItem> = useCallback(
+      ({ item: { onPress, disabled, active, icon } }) => {
+        const args = {
+          editor,
+          editorState,
+          setToolbarContext,
+          toolbarContext,
+        };
+
+        const style = createItemViewStyle({ onPress, disabled, active, icon });
+
+        return (
+          <TouchableOpacity
+            onPress={onPress(args)}
+            disabled={disabled(args)}
+            style={touchableStyle}
+            key={icon}
+          >
+            <View style={style}>
+              <Icon type={icon} />
+            </View>
+          </TouchableOpacity>
+        );
+      },
+      [
+        editor,
+        editorState,
+        setToolbarContext,
+        toolbarContext,
+        createItemViewStyle,
+        touchableStyle,
+      ]
+    );
+
+    switch (toolbarContext) {
+      case ToolbarContext.Main:
+      case ToolbarContext.Heading:
+        return (
+          <FlatList
+            data={
+              toolbarContext === ToolbarContext.Main ? items : HEADING_ITEMS
+            }
+            style={[
+              editor.theme.toolbar.toolbarBody,
+              hideToolbar ? editor.theme.toolbar.hidden : undefined,
+            ]}
+            renderItem={renderItem}
+            horizontal
+            removeClippedSubviews
+            initialNumToRender={6}
+            maxToRenderPerBatch={6}
+            getItemLayout={(_, index) => ({
```

Comment:
I'd usually say memoize this, but if the list never gets long it doesn't really matter.